external_services/api/authentication.py

```diff
@@ -0,0 +1,124 @@
+    the LTI service set for the exercise matches the oauth_consumer_key parameter
+    of the request.
+    @return tuple (boolean, error_message) boolean is True if verification succeeded,
+    False otherwise.
+    '''
+    headers = {
+        'Content-Type': request.content_type,
+        'Authorization': request.META.get('HTTP_AUTHORIZATION'),
+        'Host': request.META.get('HTTP_HOST'),
+    }
+    # all OAuth parameters must be given in the HTTP Authorization header
+    # (not in POST data or GET query parameters) when a body hash is used
+    # to secure the request body
+    all_req_oauth_params = oauthlib.oauth1.rfc5849.signature.collect_parameters(
+        headers=headers, exclude_oauth_signature=False)
+    # collect_parameters returns a list of key-value pairs
+    req_oauth_params = [] # the oauth_signature must be separated from the other parameters
+    oauth_signature = None
+    for key, value in all_req_oauth_params:
+        if key == 'oauth_signature':
+            oauth_signature = value
+        else:
+            req_oauth_params.append((key, value))
+    req_oauth_params_dict = dict(req_oauth_params)
+    
+    # check oauth_consumer_key and find the corresponding secret
+    consumer_key = req_oauth_params_dict.get('oauth_consumer_key')
+    if not consumer_key:
+        return False, 'oauth_consumer_key missing'
+    try:
+        lti_service = LTIService.objects.get(consumer_key=consumer_key)
+    except (LTIService.DoesNotExist, LTIService.MultipleObjectsReturned):
+        return False, 'unknown oauth_consumer_key'
+    
+    if lti_exercise and lti_exercise.lti_service.pk != lti_service.pk:
+        # the consumer key refers to a different LTI service than the exercise
+        return False, 'oauth_consumer_key mismatch'
+    client_secret = lti_service.consumer_secret
+    
+    # check the OAuth timestamp. Do not allow old requests in order to prevent replay attacks.
+    try:
+        timestamp = datetime.datetime.utcfromtimestamp(int(req_oauth_params_dict.get('oauth_timestamp')))
+        # oauth_timestamp: seconds since January 1, 1970 00:00:00 GMT
+    except ValueError:
+        return False, 'oauth_timestamp is missing or has an invalid format'
+    
+    now = datetime.datetime.utcnow()
+    delta = datetime.timedelta(seconds=OAuthNonceCache.CACHE_TIMEOUT_SECONDS)
+    if not (now - delta < timestamp and timestamp < now + delta):
+        return False, 'oauth_timestamp has expired'
+    
+    # check OAuth nonce: The nonce value MUST be unique across all requests with
+    # the same timestamp, client credentials, and token combinations.
+    # Previously seen nonces are kept in the cache for a few minutes
+    # (the duration must match the accepted timestamp age).
+    nonce = req_oauth_params_dict.get('oauth_nonce')
+    if not nonce:
+        return False, 'oauth_nonce missing'
+    nonce_cache = OAuthNonceCache(nonce, req_oauth_params_dict.get('oauth_timestamp'), client_secret)
```

Comment:
why `req_oauth_params_dict.get('oauth_timestamp')` and not `timestamp`?

Reply:
`req_oauth_params_dict.get('oauth_timestamp')` is a string like `"436372524"` and `timestamp` is a datetime object. I want the string format here.

Reply:
If this oauth_timestamp was rendered by us and then signed, then there shouldn't be any problem. I was afraid that non sanitized timestamp could fool nonce cache.

Reply:
The oauth_signature protects the integrity of the oauth headers. The format of oauth_timestamp is checked earlier in the code. Are you concerned that attackers could fill the cache with trash nonces even without having a valid signature?

Reply:
Nope. Just initially though it maybe could have been possible to fool the cache, but due to signature it shouldn't open any vulnerabilities.